subprojects/core/src/main/java/org/gradle/api/internal/file/archive/TarFileTree.java

```diff
@@ -206,17 +239,101 @@ public int getMode() {
         }
     }
 
-    private static class NoCloseTarInputStream extends TarInputStream {
-        public NoCloseTarInputStream(InputStream is) {
+    private static class NoCloseTarArchiveInputStream extends TarArchiveInputStream {
+        public NoCloseTarArchiveInputStream(InputStream is) {
             super(is);
         }
 
         @Override
         public void close() throws IOException {
         }
+    }
 
-        public TarEntry getCurrent() {
-            return currEntry;
+    /**
+     * Copy of <code>ArchiveStreamFactory.detect(InputStream)</code>, extended to not
+     * throw an exception for empty TAR files (ie. ones with no entries in them).
+     */
+    private static String detectFormat(InputStream inputStream) throws ArchiveException {
```

Comment:
If we need this code here then we should simplify it to only check for TAR which is the only interesting case here.
Removing the other checks will make it faster.

Reply:
Yes, I've thought about it, but had arguments against it too. For example let's say you remove the first part, which can detect other formats quickly. Then you make it potentially slower when the input is of those other formats. But, on the other hand, dealing with a non-TAR format should be a rare event, not the typical one... Yeah, should make sense to simplify. Will do it.